### PR TITLE
GC optimizations

### PR DESCRIPTION
**Pull request by roman-khimov**
Makes MPT GC even better.

### CODE REVIEW

**Comment by roman-khimov** [comment]
Win test:

> 2022-02-12T20:24:28.4848771Z     testing.go:967: TempDir RemoveAll cleanup: remove C:\Users\RUNNER~1\AppData\Local\Temp\TestAllDBs1111297819\009\test_bolt_db: Access is denied.

Really strange.

File: pkg/core/storage/boltdb_store.go [hunk 1]
```diff
@@ -110,44 +109,30 @@ func (s *BoltDBStore) PutChangeSet(puts map[string][]byte) error {
...
-		c := tx.Bucket(Bucket).Cursor()
-		// Move cursor to the first kv pair which is followed by the pair matching the specified prefix.
-		if len(start) == 0 {
-			lastKey, _ := c.Last()
-			start = lastKey
-		}
-		rng := util.BytesPrefix(start) // in fact, we only need limit based on start slice to iterate backwards starting from this limit
-		c.Seek(rng.Limit)
-		for k, v := c.Prev(); k != nil && bytes.HasPrefix(k, key); k, v = c.Prev() {
+		for ; k != nil && bytes.HasPrefix(k, rng.Prefix) && (len(rang.Limit) == 0 || bytes.Compare(k, rang.Limit) <= 0); k, v = next() {
```
Comment (AnnaShaleva) [inline pkg/core/storage/boltdb_store.go:156]: Here we can skip `bytes.HasPrefix(k, rng.Prefix)` check for non-backwards iterations to save some time, is it important?

Comment (AnnaShaleva) [reply pkg/core/storage/boltdb_store.go:156]: And `bytes.Compare(k, rang.Limit) <= 0` will always be true for backwards seeking.

File: pkg/core/storage/boltdb_store.go [hunk 2]
```diff
@@ -107,10 +107,31 @@ func (s *BoltDBStore) PutChangeSet(puts map[string][]byte) error {
 	})
 }
 
+// SeekGC implements the Store interface.
+func (s *BoltDBStore) SeekGC(rng SeekRange, keep func(k, v []byte) bool) error {
+	return boltSeek(s.db.Update, rng, func(c *bbolt.Cursor, k, v []byte) bool {
+		if !keep(k, v) {
+			if c.Delete() != nil {
+				return false
```
Comment (AnnaShaleva) [inline pkg/core/storage/boltdb_store.go:115]: Then if some key is failed to be deleted `boltSeek` will just stop iteration and return no error. Maybe we'd better return this error from `SeekGC`?

File: pkg/core/storage/store.go [hunk 3]
```diff
@@ -96,6 +96,11 @@ type (
 		// Key and value slices should not be modified.
 		// Seek can guarantee that key-value items are sorted by key in ascending way.
 		Seek(rng SeekRange, f func(k, v []byte) bool)
+		// SeekGC is similar to Seek, but the function should return true if current
+		// KV pair should be kept and false if it's to be deleted; there is no way to
+		// do an early exit here. SeekGC only works with the current Store, it won't
+		// go down to layers below and it takes a full write lock, so use it carefully.
+		SeekGC(rng SeekRange, keep func(k, v []byte) bool) error
```
Comment (AnnaShaleva) [inline pkg/core/storage/store.go:107]: And maybe it's better to rename standard `Seek(rng SeekRange, f func(k, v []byte) bool)`  function argument? I.e. to have `Seek(rng SeekRange, cont func(k, v []byte) bool)`.

File: pkg/core/stateroot/module.go [hunk 4]
```diff
@@ -152,6 +152,10 @@ func (s *Module) CleanStorage() error {
 		b.Delete(k)
 		return true
 	})
+	s.Store.Seek(storage.SeekRange{Prefix: []byte{byte(storage.DataMPTAux)}}, func(k, _ []byte) bool {
```
Comment (AnnaShaleva) [inline pkg/core/stateroot/module.go:155]: Do we remove untraceable MPT stateroots?

**Comment by AnnaShaleva** [comment]
> Win test:
> 
> > 2022-02-12T20:24:28.4848771Z     testing.go:967: TempDir RemoveAll cleanup: remove C:\Users\RUNNER~1\AppData\Local\Temp\TestAllDBs1111297819\009\test_bolt_db: Access is denied.
> 
> Really strange.

@roman-khimov, Windows is *very* sensitive about resorce releasing. Try to add `require.NoError(t, s.Close())`  in the end of `testStoreSeekGC(t *testing.T, s Store)` (just an assumption).

File: pkg/core/stateroot/module.go (hunk 4 above)
Comment (roman-khimov) [reply pkg/core/stateroot/module.go:155]: We don't, I've just reproduced the old behavior here. Maybe we can keep `MPTAux` and simplify reinstantiation below?

File: pkg/core/storage/store.go (hunk 3 above)
Comment (roman-khimov) [reply pkg/core/storage/store.go:107]: Well, we've had `f` for a long time, I think it's OK.

File: pkg/core/storage/boltdb_store.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/core/storage/boltdb_store.go:156]: I've seen that, but keeping this generic is simpler I think.

File: pkg/core/storage/boltdb_store.go (hunk 2 above)
Comment (roman-khimov) [reply pkg/core/storage/boltdb_store.go]: Fixed.

File: pkg/core/stateroot/module.go (hunk 4 above)
Comment (roman-khimov) [reply pkg/core/stateroot/module.go]: Fixed, I think we can just keep Aux, it has a root for block zero at this stage and that's exactly what we want to keep.

**Review by AnnaShaleva — COMMENTED** [review]
Minor issue is left, otherwise LGTM.

File: pkg/core/stateroot/module.go (hunk 4 above)
Comment (AnnaShaleva) [reply pkg/core/stateroot/module.go]: You're right, no need to remove it.

But the question is still relevant, currently, when `RemoveUntraceableBlocks` is on, we don't perform outdated stateroots removal, so they're stored forever. Maybe we can create a separate issue for that?

Comment (roman-khimov) [reply pkg/core/stateroot/module.go]: That's intentional, I don't think it should be fixed. They're treated like headers that we keep forever.

File: pkg/core/storage/leveldb_store.go [hunk 5]
```diff
@@ -82,13 +83,35 @@ func (s *LevelDBStore) PutChangeSet(puts map[string][]byte) error {
...
+// SeekGC implements the Store interface.
+func (s *LevelDBStore) SeekGC(rng SeekRange, keep func(k, v []byte) bool) error {
+	tx, err := s.db.OpenTransaction()
+	if err != nil {
+		return err
+	}
+	iter := tx.NewIterator(seekRangeToPrefixes(rng), nil)
+	s.seek(iter, rng.Backwards, func(k, v []byte) bool {
+		if !keep(k, v) {
+			if tx.Delete(k, nil) != nil {
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/core/storage/leveldb_store.go]: Fixed.

**Pull request merged**